#Домашнє завдання до тем apply(), groupby(), pivot_table()

В цьому домашньому завданні продовжуємо працювати з набором даних `supermarket_sales.csv`.

0. Імпортуйте бібліотеку pandas та зчитайте дані у змінну `df` типу `pandas.DataFrame`.

In [1]:
import pandas as pd
from google.colab import drive
drive.mount('/content/drive')

df = pd.read_csv('drive/MyDrive/Аналітик даних в ІТ/Data/supermarket_sales.csv')

Mounted at /content/drive


1. Дослідимо, який філіал супермаркету ('Branch') є найприбутковішим. Для цього знайдіть сумарний прибуток за кожним філіалом і виявіть, який філіал має найвищий.

In [2]:
df.head()

Invoice ID Branch       City Customer type  Gender  \
0  750-67-8428      A     Yangon        Member  Female   
1  226-31-3081      C  Naypyitaw        Normal  Female   
2  631-41-3108      A     Yangon        Normal    Male   
3  123-19-1176      A     Yangon        Member    Male   
4  373-73-7910      A     Yangon        Normal    Male   

             Product line  Unit price  Quantity   Tax 5%     Total       Date  \
0       Health and beauty       74.69         7  26.1415  548.9715   1/5/2019   
1  Electronic accessories       15.28         5   3.8200   80.2200   3/8/2019   
2      Home and lifestyle       46.33         7  16.2155  340.5255   3/3/2019   
3       Health and beauty       58.22         8  23.2880  489.0480  1/27/2019   
4       Sports and travel       86.31         7  30.2085  634.3785   2/8/2019   

    Time      Payment    cogs  gross margin percentage  gross income  Rating  
0  13:08      Ewallet  522.83                 4.761905       26.1415     9.1  
1  10:29         Cash   76.40                 4.761905        3.8200     9.6  
2  13:23  Credit card  324.31                 4.761905       16.2155     7.4  
3  20:33      Ewallet  465.76                 4.761905       23.2880     8.4  
4  10:37      Ewallet  604.17                 4.761905       30.2085     5.3

In [3]:
df.groupby('Branch')['gross income'].sum()

Branch
A    5057.1605
B    5057.0320
C    5265.1765
Name: gross income, dtype: float64

In [5]:
df.groupby(['Branch','City'])['gross income'].sum()

Branch  City     
A       Yangon       5057.1605
B       Mandalay     5057.0320
C       Naypyitaw    5265.1765
Name: gross income, dtype: float64

**Опис:**

Спочатку знайшла сумарний валовий прибуток в розрізі філій.
Але для себе треба однозначно
 - перевірити, чи кожна філія відповідає одному місту;
 - переконатися, що філії не дублюються між містами
Для цього групування додатково зроблене, як в розрізі філій, так і в розрізі міст

**Висновок:**

*Найприбутковішим філіалом є Branch C в місті Naypyitaw.
Оскільки 5265.18 > 5057.16 > 5057.03.*

2. В якому місті знайходиться філіал з найвищим прибутком? Може в тому місці нам розмітисти ще один магазин.  
Знайдіть відповідь за допомогою функціоналу Pandas.

In [6]:
# в попередньому завданні дали відповідь на питання цього завдання

df.groupby(['Branch','City'])['gross income'].sum()

Branch  City     
A       Yangon       5057.1605
B       Mandalay     5057.0320
C       Naypyitaw    5265.1765
Name: gross income, dtype: float64

**Висновок:**

Місто Naypyitaw має найбільший валовий прибуток  в 5265.18. Це може говорити про
 - більший попит населення міста,
 - більшу купівельну спроможність або
 - більш ефективну роботу керівництва та персоналу філіалу.

*Логічно розглянути можливість відкриття додаткового магазину в цьому місті, але перед рішенням варто перевірити інші метрики — середній чек, кількість транзакцій, попит на товарні категорії та конкуренцію*

3.1. Створіть зводну таблицю, яка покаже, скільки покупок (інвойсів) було зроблено в кожній з філій (`Branch`) за різними категоріями товарів. Запишіть таблицю в змінну `invoices_by_category` і виведіть змінну на екран.
Ця таблиця допоможе проаналізувати, в якій філії купують найбільше товарів кожної з категорій.

In [15]:
invoices_by_category = pd.pivot_table(df, index = 'Branch',
                                      columns = ['Product line'],
                                      values = ['Invoice ID'],
                                      aggfunc = "count")
invoices_by_category.columns = invoices_by_category.columns.droplevel(0)    # це прибере мультііндекс і рівень стовпців ['Invoice ID'] без значень
                                                                            # і спростить структуру таблиці
invoices_by_category


Product line  Electronic accessories  Fashion accessories  Food and beverages  \
Branch                                                                          
A                                 60                   51                  58   
B                                 55                   62                  50   
C                                 55                   65                  66   

Product line  Health and beauty  Home and lifestyle  Sports and travel  
Branch                                                                  
A                            47                  65                 59  
B                            53                  50                 62  
C                            52                  45                 45

Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1rueAdko6S3UxIHGtojetTxlES-EyM6Yb)

3.2. Викристовуючи змінну `invoices_by_category` дайте відповідь програмно (тобто значення треба не просто знайти очима, а вивести за допомогою коду), в якому філіалі магазину (`Branch`) найбільше інвойсів із покупкою товарів категорії "Електронні аксесуари" (`Electronic accessories`)?


In [21]:
top_branch = invoices_by_category['Electronic accessories'].idxmax()
max_invoices = invoices_by_category['Electronic accessories'].max()

print(f'В філіалі {top_branch} найбільша кількість інвойсів у категорії "Електронні аксесуари" - {max_invoices}')

В філіалі A найбільша кількість інвойсів у категорії "Електронні аксесуари" - 60


4-6. **Творче завдання на розвиток аналітичного мислення**

Крок 1. Сформулюйте ТРИ питання (гіпотези) до наявних даних, які допомогли б вам зрозуміти, які користувачі що, де та коли найбільше/найменше купують, аби дати на основі цих гіпотез рекомендації бізнесу. Звісно питання мають бути не тими, на які ми вже відповіли в завданнях модулю.

Крок 2. Знайдіть відповіді на свої питання з допомогою функціоналу pandas.

Крок 3. Напишіть, як відповідь на це питання може бути використана для прийняття бізнес рішень.   
   
 Питання можуть бути будь-якої складності, але їх має бути 3. Кожне питання оцінюється як 1 завдання. Без виконання цього завдання ДЗ не приймається. Якщо є питання щодо виконання - пишіть у чат 🙌


**По-перше**,

*спробуємо проаналізувати, а які клієнти (чоловіки, жінки) приносять найбільший прибуток в розрізі міст філій*

**та відповісти на перше питання**,
*який внесок у прибуток роблять різні групи покупців у кожному місті?*

In [51]:
gender_sales_cities = pd.pivot_table(df,
                                    index='City',
                                    columns='Gender',
                                    values='gross income',
                                    aggfunc='sum'
                                    ).round(2)
gender_totals = gender_sales_cities.sum()                             # Загальний прибуток для Female та Male
gender_part = (gender_totals / gender_totals.sum() * 100).round(0)    # Частка (%) від загального прибутку

gender_sales_cities.loc['Part'] = gender_part
gender_sales_cities


Gender      Female     Male
City                       
Mandalay   2520.40  2536.64
Naypyitaw  2937.40  2327.77
Yangon     2536.63  2520.53
Part         52.00    48.00

**Висновок 1:**

*За аналізуємий період жінки принесли більше половини прибутку мережі. Сумарний прибуток від жінок становить 52%, від чоловіків - 48%. Різниця невелика, але стабільна в межах міст Naypyitaw та Yangon, що означає системну поведінку клієнтів, а не випадковість.*

**Бізнес-рекомендації:**

1. Збільшити прибуток за рахунок найбільш прибуткової групи (жінки) (підсилити присутність товарів, які частіше купують жінки, запустити банери/промо саме на жіночі категорії (health & beauty, fashion accessories тощо).
2. Збільшити середній чек серед чоловіків (акції на "чоловічі" категорії — Electronic accessories, Home & lifestyle, Sports & travel, цільові кампанії в соцмережах (Facebook/Instagram)
3. Покращити пропозиції за містами (локальний маркетинг) (у Yangon і Naypyitaw можна посилити жіночу аудиторію, адже там вони вже лідирують; у Mandalay варто більше працювати з чоловічою аудиторією)

**По-друге,**

*варто проаналізувати, чи карткові покупки мають вищий чек і можливо, варто робити промо банківських карток; чи cash має менший чек і можливо, треба стимулювати електронну оплату; чи різні міста платять по-різному?*

**Тому наступне питання, на яке варто відповісти:**

*чи відрізняється середній чек між способами оплати і чи це залежить від міста?*


In [53]:
cities_payment = pd.pivot_table(df,
                               index='City',
                               columns='Payment',
                               values='Total',
                               aggfunc='mean'
                               ).round(2)
cities_payment

Payment      Cash  Credit card  Ewallet
City                                   
Mandalay   321.27       342.61   296.58
Naypyitaw  347.47       309.46   350.52
Yangon     307.10       318.22   312.10

**Висновок:**

В Mandalay клієнти з кредитними картками купують більше, ніж ті, хто платить готівкою чи електронним гаманцем.
В Naypyitaw клієнти охочіше платять великими сумами електронними гаманцями.
В Yangon середній чек при оплаті карткою стабільно більший.

Таким чином, в двох містах із трьох (Yangon, Mandalay) найбільші покупки роблять власники кредитних карток. Готівка в усіх містах має другий середній чек або найменший

**Бізнес-рекомендації:**

1. Стимулювати оплату карткою у Yangon та Mandalay (розіграш товарів тільки серед покупців із картками, пріоритетна каса)
2. Розвивати та просувати E-wallet у Naypyitaw (найвищий середній чек) (додати бонусні бали за оплату e-wallet, зробити QR-оплату швидшою, зробити промо)
3. Стимулювання перехід з Cash на інші типи оплат

**Тому що дані показують, що тип оплати прямо впливає на розмір чеку, на прибуток магазинів!!!**

**Третє питання:**

у яких годинах дня прибуток максимальний і чи співпадає це з піковими годинами покупок?
Це дозволить:
- оптимізувати графік персоналу (каси чи склади);
- визначити години для акцій, промо чи дегустацій;
- проаналізувати, у пікові години люди купують менше дорогих товарів і чи потрібна зміна викладки полиць?

In [65]:
def get_hour(t):
    return int(t.split(':')[0])

df['Hour'] = df['Time'].apply(get_hour)

# сумарний прибуток по годинам
income_per_hour = df.groupby('Hour')['gross income'].sum().round(2)
income_per_hour


Hour
10    1496.26
11    1446.54
12    1241.23
13    1653.49
14    1468.02
15    1484.74
16    1201.25
17    1164.06
18    1239.54
19    1890.45
20    1093.79
Name: gross income, dtype: float64

In [60]:
# кількість інвойсів по годинам (пікові години)
transactions_per_hour = df.groupby('Hour')['Invoice ID'].count()
transactions_per_hour

Hour
10    101
11     90
12     89
13    103
14     83
15    102
16     77
17     74
18     93
19    113
20     75
Name: Invoice ID, dtype: int64

In [66]:
# Об'єднана таблиця
hour_analysis = pd.DataFrame({
                            'trans_count': transactions_per_hour,
                            'gross_income_sum': income_per_hour
                            })
hour_analysis = hour_analysis.sort_values(by='gross_income_sum', ascending=False)
hour_analysis

trans_count  gross_income_sum
Hour                               
19            113           1890.45
13            103           1653.49
10            101           1496.26
15            102           1484.74
14             83           1468.02
11             90           1446.54
12             89           1241.23
18             93           1239.54
16             77           1201.25
17             74           1164.06
20             75           1093.79

**Висновки**:

Найприбутковіші години — 19:00, 13:00, 10:00, де
1. 19:00 — абсолютний лідер як за кількістю покупок - 113, так і за сумою прибутку - 1890.45, це період після роботи, коли клієнти роблять великі закупки.
2. 13:00 та 15:00 — пікові години обіду, люди приходять за швидкими покупками та перекусами, але загальний прибуток тут дуже значний.
3. 10:00 — ранкове відкриття магазинів - свіжі покупки, покупка необхідних товарів на день чи закупівля наперед.

**Бізнес-рекомендації:**

1. Оптимізація графіка персоналу. Враховуючи піковість 10:00, 13:00, 15:00, 19:00, варто:
- збільшити кількість персоналу на касах у ці години
- підсилити викладку товару перед початком пікового періоду;
- оптимізувати графік розвантаження поставок, щоб не перетиналося з піковими годинами
2. Розміщення промо та акцій у правильні години
